test/unit/assorted/server_selection_logic_spec_utils.ts

```diff
@@ -0,0 +1,151 @@
+  descriptions: ServerDescription[]
+): Map<string, ServerDescription> {
+  const descriptionMap = new Map<string, ServerDescription>();
+
+  for (const description of descriptions) {
+    descriptionMap.set(description.address, description);
+  }
+
+  return descriptionMap;
+}
+
+/**
+ * Executes a server selection logic test
+ * @see https://github.com/mongodb/specifications/tree/master/source/server-selection/tests#server-selection-logic-tests
+ */
+export function runServerSelectionLogicTest(testDefinition: Test) {
+  const allHosts = testDefinition.topology_description.servers.map(({ address }) => address);
+  const serversInTopology = testDefinition.topology_description.servers.map(s =>
+    serverDescriptionFromDefinition(s, allHosts)
+  );
+  const serverDescriptions = serverDescriptionsToMap(serversInTopology);
+  const topologyDescription = new TopologyDescription(
+    testDefinition.topology_description.type,
+    serverDescriptions
+  );
+  const expectedServers = serverDescriptionsToMap(
+    testDefinition.in_latency_window.map(s => serverDescriptionFromDefinition(s))
+  );
+
+  let selector;
+  if (testDefinition.operation === 'write') {
+    selector = ServerSelectors.writableServerSelector();
+  } else if (testDefinition.operation === 'read' || testDefinition.read_preference) {
+    try {
+      const readPreference = readPreferenceFromDefinition(testDefinition.read_preference);
+      selector = ServerSelectors.readPreferenceServerSelector(readPreference);
+    } catch (e) {
+      expect(e, ejson`Invalid readPreference: ${testDefinition.read_preference}`).not.to.exist;
+    }
+  }
+
+  const result = selector(topologyDescription, serversInTopology);
+
+  expect(result.length).to.equal(expectedServers.size);
+
+  for (const server of result) {
+    const expectedServer = expectedServers.get(server.address);
+    expect(expectedServer).to.exist;
+    compareServerDescriptions(server, expectedServer);
+    expectedServers.delete(server.address);
+  }
+
+  expect(expectedServers.size).to.equal(0);
+}
+
+export function collectServerSelectionLogicTests(specDir) {
+  const testTypes = readdirSync(specDir).filter(d => statSync(join(specDir, d)).isDirectory());
+
+  const tests = {};
+  for (const testType of testTypes) {
+    const testsOfType = readdirSync(join(specDir, testType)).filter(d =>
+      statSync(join(specDir, testType, d)).isDirectory()
+    );
+    const result = {};
+    for (const subType of testsOfType) {
+      result[subType] = readdirSync(join(specDir, testType, subType))
+        .filter(f => extname(f) === '.json')
+        .map(f => {
+          const fileContents = readFileSync(join(specDir, testType, subType, f), {
+            encoding: 'utf-8'
+          });
+          const test = EJSON.parse(fileContents, { relaxed: true }) as unknown as Document;
+          test.name = basename(f, '.json');
+          test.type = testType;
+          test.subType = subType;
+          return test;
```

Comment:
Similar to what I did [here](https://github.com/mongodb/node-mongodb-native/blob/main/test/unit/assorted/server_discovery_and_monitoring.spec.test.ts#L149-L171) for the SDAM runner can we assert what we know to be true about the test shape at this stage, it gives us a warning for future syncs if there's something wrong with the format. **Or** if there's a new key that we don't have documented in our types, which means we probably don't test for it

Reply:
I see how the tests cover this themselves, so no need for an input validator.